File: Perspectives/src/d3/GraphD3Viewer.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package d3;

import Graph.Graph;
import Graph.GraphData;
import java.awt.Color;
import java.util.ArrayList;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import properties.PColor;
import properties.Property;

/**
 *
 * @author rajin
 */
public class GraphD3Viewer extends D3Viewer {
    private final String PROPERTY_NODE_COLOR ="Node Color";
    private final String PROPERTY_LINK_COLOR ="Link Color";
    private Graph graph;
    public GraphD3Viewer(String name, GraphData graphData)            
    {
        super(name);
        this.graph = graphData.graph;
        try {

           Property<PColor> p1 = new Property<PColor>(PROPERTY_NODE_COLOR,new PColor(new Color(0, 255, 0)));
          
            this.addProperty(p1);
            
             Property<PColor> p2 = new Property<PColor>(PROPERTY_LINK_COLOR,new PColor(new Color(0, 0, 0)));
          
            this.addProperty(p2);

           

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    private Color getColor(String propertyName)
    {
        Property<PColor> p = this.getProperty(propertyName);
        return p.getValue().colorValue();
    }
    private String getHexString(Color color)
    {
        String hex = String.format("#%02x%02x%02x",color.getRed(), color.getGreen(), color.getBlue());
        return hex;
    }

    @Override
    public JSONObject updateData(boolean isInitialCall)
    {
        JSONObject data = new JSONObject();
        //InitialCall
        data.put("IsInitialCall", isInitialCall);
        // Property Data
        JSONObject propertyData= new JSONObject();
        
        String nodeColor = this.getHexString(this.getColor(PROPERTY_NODE_COLOR));
        propertyData.put("NodeColor", nodeColor);
        
        String edgeColor = this.getHexString(this.getColor(PROPERTY_LINK_COLOR));
        propertyData.put("LinkColor", edgeColor);
        
        data.put("PropertyData", propertyData);
        //Graph Data
        JSONObject graphJSON = new JSONObject();
        ArrayList<String> nodes = this.graph.getNodes();
        ArrayList<Integer> e1 = new ArrayList<Integer>();
        ArrayList<Integer> e2 = new ArrayList<Integer>();
        this.graph.getEdgesAsIndeces(e1, e2);
        
        JSONArray nodeArray = new JSONArray();
        for(String node: nodes)
        {
            JSONObject obj = new JSONObject();
            obj.put("name", node);
            nodeArray.add(obj);
        }
        graphJSON.put("nodes", nodeArray);
        
        JSONArray linkArray = new JSONArray();
        for(int i=0;i<e1.size();i++)
        {
            int source = e1.get(i).intValue();
            int destination = e2.get(i).intValue();
            JSONObject obj = new JSONObject();
            obj.put("source", source);
            obj.put("target", destination);
            linkArray.add(obj); 
        }
        graphJSON.put("links", linkArray);
        
        data.put("graph", graphJSON);
        return data;
    }
   
}

File: VizOnline_Web/web/d3/JSClients/GraphD3Viewer.js
/* 
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */


function JSD3Client()
{ 
};

JSD3Client.prototype.dataUpdated = function (data, canvas, canvasWidth, canvasHeight)
{
    if(data.IsInitialCall)
    {
        init(data, canvas, canvasWidth, canvasHeight);
    }
    else
    {
        update(data, canvas);
    }
        
    

}
function init(data, canvas, canvasWidth, canvasHeight)
{
    var graph = data.graph;
    var color = d3.scale.category20();
    console.log(JSON.stringify(color));

    var force = d3.layout.force()
        .charge(-120)
        .linkDistance(30)
        .size([canvasWidth, canvasHeight]);

    var svg = d3.select(canvas).append("svg")
        .attr("width", canvasWidth)
        .attr("height", canvasHeight);


      force
          .nodes(graph.nodes)
          .links(graph.links)
          .start();
    layout = force;
      var link = svg.selectAll(".link")
          .data(graph.links)
        .enter().append("line")
          .attr("class", "link")
          .style("stroke-width","1px")
          .style("stroke",  data.PropertyData.LinkColor);

      var node = svg.selectAll(".node")
          .data(graph.nodes)
        .enter().append("circle")
          .attr("class", "node")
          .attr("r", 5)
          .style("fill", data.PropertyData.NodeColor)
          .call(force.drag);
    //
      node.append("title")
          .text(function(d) { return d.name; });

      force.on("tick", function() {
        link.attr("x1", function(d) {
                            return d.source.x; 
                        })
            .attr("y1", function(d) { return d.source.y; })
            .attr("x2", function(d) { return d.target.x; })
            .attr("y2", function(d) { return d.target.y; });

        node.attr("cx", function(d) { return d.x; })
            .attr("cy", function(d) { return d.y; });
      });
}


function update(data, canvas)
{
    d3.select(canvas).selectAll(".node")
          .style("fill", data.PropertyData.NodeColor);
    d3.select(canvas).selectAll(".link")
          .style("stroke",  data.PropertyData.LinkColor);
}